test/integration/lidar_visual.cc

```diff
@@ -0,0 +1,572 @@
+  // listen to new gpu rays frames
+  float *scan2 = new float[hRayCount * vRayCount * 3];
+
+  gpuRays2->Update();
+  // Test Copy method instead of using the callback for the second rays caster
+  gpuRays2->Copy(scan2);
+
+  // copy ray caster values to lidar data
+  std::vector<double> pts2;
+  for (int j = 0; j < vRayCount; j++)
+  {
+    for (int i = 0; i < gpuRays->RayCount(); ++i)
+    {
+      pts2.push_back(scan2[j*channels*gpuRays->RayCount() + i * channels]);
+    }
+  }
+  lidarVis2->SetPoints(pts2);
+  lidarVis2->Update();
+  std::vector<double> pts_back2 = lidarVis2->Points();
+
+  // Only box01 should be visible to rays caster 2
+  EXPECT_DOUBLE_EQ(pts_back2[0], maxRange);
+  EXPECT_NEAR(pts_back2[mid], expectedRangeAtMidPointBox1, LASER_TOL);
+  EXPECT_DOUBLE_EQ(pts_back2[last], maxRange);
+
+  // Move all boxes out of range
+  visualBox1->SetWorldPosition(
+      ignition::math::Vector3d(maxRange + 1, 0, 0));
+  visualBox1->SetWorldRotation(box01Pose.Rot());
+  visualBox2->SetWorldPosition(
+      ignition::math::Vector3d(0, -(maxRange + 1), 0));
+  visualBox2->SetWorldRotation(box02Pose.Rot());
+
+  gpuRays->Update();
+  gpuRays2->Update();
+  gpuRays2->Copy(scan2);
+
+  pts.clear();
+  for (int j = 0; j < vRayCount; j++)
+  {
+    for (int i = 0; i < gpuRays->RayCount(); ++i)
+    {
+      pts.push_back(scan[j*channels*gpuRays->RayCount() + i * channels]);
+    }
+  }
+
+  pts2.clear();
+  for (int j = 0; j < vRayCount; j++)
+  {
+    for (int i = 0; i < gpuRays->RayCount(); ++i)
+    {
+      pts2.push_back(scan2[j*channels*gpuRays->RayCount() + i * channels]);
+    }
+  }
+
+  for (unsigned int i = 0; i < lidarVis->HorizontalRayCount(); ++i)
+    EXPECT_DOUBLE_EQ(pts[i], ignition::math::INF_D);
+
+  for (unsigned int i = 0; i < lidarVis->HorizontalRayCount(); ++i)
+    EXPECT_DOUBLE_EQ(pts2[i], maxRange);
+
+  c.reset();
+
+  delete [] scan;
+  delete [] scan2;
+
+  scan = nullptr;
+  scan2 = nullptr;
+
+  // Clean up
+  engine->DestroyScene(scene);
+  rendering::unloadEngine(engine->Name());
+}
+
+/////////////////////////////////////////////////
+/// \brief Test GPU rays vertical component
+void LidarVisualTest::LaserVertical(const std::string &_renderEngine)
+{
+#ifdef __APPLE__
+  std::cerr << "Skipping test for apple, see issue #35." << std::endl;
+  return;
+#endif
+
+  if (_renderEngine == "optix")
+  {
+    igndbg << "GpuRays not supported yet in rendering engine: "
+            << _renderEngine << std::endl;
+    return;
+  }
+
+  // Test a rays that has a vertical range component.
+  // Place a box within range and verify range values,
+  // then move the box out of range and verify range values
+
+  double hMinAngle = -IGN_PI/2.0;
+  double hMaxAngle = IGN_PI/2.0;
+  double vMinAngle = -IGN_PI/4.0;
+  double vMaxAngle = IGN_PI/4.0;
+  double minRange = 0.1;
+  double maxRange = 5.0;
+  unsigned int hRayCount = 640;
+  unsigned int vRayCount = 4;
+
+  common::Time waitTime = common::Time(WAIT_TIME);
```

Comment:
You probably got this from `gpu_rays` test but I think it is not used. We can remove this